raccine.cpp

```diff
@@ -249,6 +260,12 @@ int wmain(int argc, WCHAR* argv[]) {
         else if (_wcsicmp(L"ignoreallfailures", argv[iCount]) == 0) {
             bIgnoreallFailures = true;
         }
+	else if (convertedArg.find(L"win32_shadowcopy") != std::string::npos || convertedArg.find(L"Win32_Shadowcopy") != std::string::npos) {
+            bwin32ShadowCopy = true;
+        }
+        else if (_wcsicmp(L"-encodedcommand", argv[iCount]) == 0) {
```

Comment:
The Emotet command line typically looks like 
```
-enc JAB..
-encod JAB..
-encoded JAB..
```

Do you know how to access the `JAB...` Parameter value?

We would have to check, 
Does the Parameter starts with `-e` followed by a value that starts with `JAB`. 
Is that possible? I myself though that we could use a substring of the first 2 characters to match against `-e`. But I don't know, without testing, how to access the `JAB`. Ist it the next parameter?`
Then we could create a condition that uses argv[iCount] and argv[iCount+1].

Reply:
I believe it's possible. It is the next parameter.

You could take the wide string version of the argv[iCount] I created above, for re-use.
This allows for using .find on a string, which works similar to .contains in other languages.

wchar_t* convertedCh = argv[iCount];
std::wstring convertedArg(convertedCh);

So, for '-e' you can use _wcsicmp, but for other arguments you can use the condition:
convertedArg.find(L"win32_shadowcopy") != std::string::npos

If true convertedArg contains "win32_shadowcopy" string.

So the boolian logic of argv[iCount] and argv[iCount+1] ... still works.

Reply:
Where does the `win32_shadowcopy` come from? 
Other methods that use Powershell to delete shadow copies. 
Your idea is to cover methods used by Ransomware with PowerShell commands in the same step as we block Emotet infections, right?

Reply:
Ah, sorry, I missed your example in the description 
```
powershell -command "Get-WmiObject Win32_Shadowcopy | ForEach-Object {$_.Delete();}"
```

Reply:
Something like this.

bool bEnc = false;
 bool bJab = false;
...
else if (convertedArg.find(L"-enc") != std::string::npos || convertedArg.find(L"-ENC") != std::string::npos) {
            bEnc = true;
    }
else if (convertedArg.find(L"jab") != std::string::npos || convertedArg.find(L"jab") != std::string::npos) {
            bJab = true;
    }
...
	(bPowerShell && bEnc && bJab)

Reply:
. find() is case sensitive, unlike _wcsicmp, so both lower case and regular are needed I believe.

Reply:
Works in first tests.
If you want I can commit the patch for:

`-enc JAB..
-encod JAB..
-encoded JAB..`

Or we can leave it out.

Reply:
It never occured to me, but couldn't we make the comparisons always against the lowercase casing? 
In e.g. Python I'd always compare `"powershell" == value.lower()` to make it a case-insensitive comparison. It wouldn't hurt in the other cases neither.

Reply:
I'd do something like :

```
value = argv[iCount].lower() 
```

and compare always with lower case strings in all comparisons.

Reply:
I'd like to see this expression
```
value = argv[iCount].lower() 
...
else if (_wcsicmp(L"-e", value.substring(0,2) == 0 && _wcsicmp(L"JAB", argv[iCount+1].substring(0,3) ==) {
```
but you have to help me with the C++ syntax

Reply:
Absolutely. I wasn't even thinking about it.
We can do:

`transform(str.begin(), str.end(), str.begin(), ::tolower)`

I will definitely implement this.

Why not lowering every arg at the start of the loop, and compare it then always with lowercase commands? 

Why skip JAB? It will become 'jab' with the transform?

Reply:
Yes, let's compare every argument with the lower-case form, except JAB. 
The `JAB` is the beginning of a Base64 encoded statement and the lower-case `jab` has a different meaning when decoded. 

See: 
https://gchq.github.io/CyberChef/#recipe=From_Base64('A-Za-z0-9%2B/%3D',true)Decode_text('UTF-16LE%20(1200)')&input=SkFCR0FEVUFhUUJpQUhjQU1BQmpBRDBBS0FBbkFGSUFKd0FyQUNnQUp3QTJBQ2NBS3dBbg

Reply:
`JAB` stands for a `$` sign in wide formatted code blocks. I once mage a cheat sheet for the most common expressions that we find in malware. 

https://gist.github.com/Neo23x0/6af876ee72b51676c82a2db8d2cd3639

Reply:
My guess is that we could be inclined to add `SUVY` and `SQBFAF` at some point in the future.

Reply:
Nice. I didn't even know the use of JAB vs jab  :)

The last commit should contain the fixes for the functionality we talked about. Please, take a look.

https://github.com/Neo23x0/Raccine/pull/17/commits/075aa40e475144a4a0d39e9599a71e54f70a1011

Reply:
SUVY  and SQBFAF. Definitely. I do have some "umbrella" ideas for those. 

Also considered data exfiltration. It's one thing not to get ransomed, but the leakage still may occur. Especially, if the victim doesn't pay, it's usually released to the public. Idea would be to look for stuff like this:

`powershell.exe -File "C:\my_path\yada_yada\run_import_script.ps1"` 
`powershell.exe -executionpolicy bypass ".\myscript.ps1" yadayadayada` 
`powershell.exe Unblock-File -Path '.\Start-ActivityTracker.ps1'`
`powershell.exe -command "Copy-Item -Path 'C:\PointA\260.txt' -Destination C:\PointB\"`

Reply:
I'm also considering this simple method that goes straight from elevated cmd. Deletes backup files in a manner similar to Ryuk ransomware.

`del /s /f /q c:\*.VHD c:\*.bac c:\*.bak c:\*.wbcat c:\*.bkf c:\Backup*.* c:\backup*.* c:\*.set c:\*.win c:\*.dsk`

Now we are talking CMD injection :)

Reply:
There is a lot of legitimate powershell commands that has strange command lines. 
However, the JAB and SUVY or SQBFAF has very low false positive rates. 

The "del" would be difficult since it isn't a standalone program but a function within cmd.exe.

Reply:
Agreed.

Reply:
Ah, I made a mistake. I have to compare the JAB to the next argument. Have to fix that.

Reply:
Looks good.
Yea, you still need the bool logic for 2 parameters detected together: '-e' and one in 'encodedCommands' list.

Reply:
I fixed it ... should work now
![Screenshot 2020-10-15 185142](https://user-images.githubusercontent.com/2851492/96161372-80881b00-0f17-11eb-98aa-be0edee4d8a4.png)

Reply:
> you still need the bool logic for 2 parameters detected together: '-e' and one in 'encodedCommands' list.

This should work via the comparison with the previous argument in the moment of the check, the current argument starts with e.g. `JAB` and the previous one with `-e`. 

I started the loop at argument number 1 instead of 0 to avoid an underflow. 
The first argument should always be the program, e.g. "Powershell.exe", doesn't it?

Reply:
0 is the source - in our case it will be 'Raccine.exe' 
1 is then then 'powershel'l or 'powershell.exe'

Reply:
Ah, okay, so argv[0] isn't interesting at all. 

Do you have an idea why it selects PID 0 for a kill. (see my screenshot)
It shouldn't.

![Screenshot 2020-10-15 190839](https://user-images.githubusercontent.com/2851492/96163190-dd84d080-0f19-11eb-83ac-2c341b656319.png)

Reply:
Hmm, that's pseudo-process, but PID=0 is the response you get when querying the WMI and some APIs, for some reason. Not sure how documented this is at all. The last time I checked it wasn't.

Reply:
I wonder how it gets to the line 
```
wprintf(L"Collecting PID %d for a kill\n", pid);
```
when `pid` is `0`. 
It clearly prints out `Collecting PID 0 for a kill`

Reply:
Ah, holy ... there's an issue
![Screenshot 2020-10-15 192721](https://user-images.githubusercontent.com/2851492/96165104-7583b980-0f1c-11eb-9f06-39ad9da1be3a.png)

Reply:
Now everything is fine
![Screenshot 2020-10-15 192941](https://user-images.githubusercontent.com/2851492/96165399-dad7aa80-0f1c-11eb-80f5-dc9b22fdd7c9.png)

Reply:
Looks like a copy-paste mistake to me :) duplicate by accident.

Reply:
Doesn't intercept lower case `jab` ... perfect 

![Screenshot 2020-10-15 193059](https://user-images.githubusercontent.com/2851492/96165489-f9d63c80-0f1c-11eb-9545-ca2fa40c854e.png)

Reply:
I thought it was weird, but since it worked (bcs PID0 is not a real process) there is no error. 
I'm over here building the detection log :)
Kill it with fire! 🗡️

Reply:
I can confirm all cases work as intended.